### DISCUSSION TITLE
Sometimes it doesn't work as expected, wrong use/configuration related to thread safety?

### DISCUSSION THREAD

**Discussion by fedxyz** [vladimir-bukhtoyarov/bucket4j#515] [Q&A] [answered]
Hi,

I have a class where I configured
```
    private static final ProxyManager<String> buckets = new CaffeineProxyManager<>(
            Caffeine.newBuilder(),
            Duration.ofMinutes(30));

    private static final BucketConfiguration bucketConfig = BucketConfiguration.builder().
            addLimit(Bandwidth.builder().capacity(1).refillIntervally(1, Duration.ofSeconds(10)).build()).
            build();

    private static final ExecutorService executor = Executors.newCachedThreadPool();
```

then I use the `bucket` in another part of my code in this way:

```
           executor.submit(() -> {

               Bucket bucket = buckets.builder().build(key,
                       () -> bucketConfig);
                    
               if(bucket.tryConsume(1)){
                        
                   // do stuff
                }
```

In the tests the `key` is always the same key, so I suppose I am using the same `bucket` every time.

From what I understand for this configuration it should permit me to consume a token every 10 seconds.
It works most of the times but sometimes this is not true, I can consume the token before 10 seconds have passed from the last used token (when `tryConsume(1)` returned true).

I am sorry I can't give a full test case but it's so random that I don't know where it can be the problem, I suppose my code is not thread safe or there is a configuration error?

Thanks for the help.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
> From what I understand for this configuration it should permit me to consume a token every 10 seconds.

Definitely no. Token bucket does not allow to configure limit `N tokens per M time window`. See more https://en.wikipedia.org/wiki/Token_bucket

Two tokens(for example in 3 sec time window) can be consumed in following scenario:
* Bucket had been created at time T0 and 1 token was consumed at T0, available_tokens becames to 0.
* Then at time `T0 + 18` sec you are tying to consume 1 token. Firstly 1 token becomes refilled and  available_tokens became to one, and because you are selected intervally refill(instead of gready) last_refill_time becomes `T0+10`(instead of `T0+18` if gready refill is chosen). Then requested 1 token is being consumed, and available tokens becomes to 0.
* Then at time `T0+21` you are trying to consume yet another token. Because current_time - available_tokens = 11sec(that greater that 10) one token is beeing refilled again, and available_tokens becoames to 1, and last_refill_time becomes to 20, you successfully consume 1 token and available_tokens became to 0.
* What do we have? 1 token was consumed at  `T0 + 18`  and 1 token was consumed at time `T0 + 21`, so 2 tokens were consumed at 3 seconds.

To avoid scenario above use greedy refill, at least you will be protected from consumption 2 tokens in time less then 10seconds. However you should undestand that in 10second it is possible to consume 2 tokens, because 1 token is able to be consumed at bucket create time, and 1 token will be refilled in 10 seconds.
